### PR TITLE
feat: style mammotome forms

### LINKED ISSUE #208: Style mammotome forms

### PR DESCRIPTION
**Pull request by Pareesh**
Completed: 
https://www.mammotome.com/en/submit-a-tech-support-ticket
https://www.mammotome.com/contact/innovate-with-mammotome/
https://www.mammotome.com/contact/investigator-initiated-study-request/
https://www.mammotome.com/contact/
https://www.mammotome.com/schedule-a-demo/

Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #208 

Test URLs:
- Before: https://main--mammotome--hlxsites.hlx.page/us/en/contact/
- After: https://mammotomeformstyle--mammotome--hlxsites.hlx.page/us/en/contact/
- Before: https://main--mammotome--hlxsites.hlx.page/us/en/schedule-a-demo
- After: https://mammotomeformstyle--mammotome--hlxsites.hlx.page/us/en/schedule-a-demo
- Before: https://main--mammotome--hlxsites.hlx.page/us/en/submit-a-tech-support-ticket
- After: https://mammotomeformstyle--mammotome--hlxsites.hlx.page/us/en/submit-a-tech-support-ticket
- Before: https://main--mammotome--hlxsites.hlx.page/us/en/about-us/innovate-with-mammotome
- After: https://mammotomeformstyle--mammotome--hlxsites.hlx.page/us/en/about-us/innovate-with-mammotome
- Before: https://main--mammotome--hlxsites.hlx.page/us/en/about-us/investigator-initiated-study-request
- After: https://mammotomeformstyle--mammotome--hlxsites.hlx.page/us/en/about-us/investigator-initiated-study-request

### CODE REVIEW

File: blocks/form/file.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+function attachFiles(wrapper, fileList, fileTemplate) {
+    for(let i = 0; i < fileList.length; i++) {
+        const file = fileList[i];
```
Comment (solaris007) [inline blocks/form/file.js:3]: you could check allowed file extensions and mime types as a basic spoofing protection in the following line:

```
        if (!isFileAllowed(file)) {
            console.error(`File extension or mime type not allowed: ${file.name}`);
            continue;
        }
```

the function could look like this:
```
function isFileAllowed(file) {
  const allowedExtensions = ['jpg', 'jpeg', 'png', 'gif', 'pdf']; // Customize this list based on your needs
  const allowedMimeTypes = ['image/jpeg', 'image/png', 'image/gif', 'application/pdf']; // Customize this list based on your needs

  const extension = file.name.split('.').pop().toLowerCase();
  const mimeType = file.type; // The MIME type can be accessed directly from the file object in the browser

  return allowedExtensions.includes(extension) && allowedMimeTypes.includes(mimeType);
}

```

File: blocks/form/file.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        const input = fileTemplate.cloneNode(true);
+        input.id = input.name = 'attachment_' + Date.now() + '' + Math.floor((Math.random() * 1000));
```
Comment (solaris007) [inline blocks/form/file.js:5]: you could reduce the risk of collisions even more by  converting to a base-36 string:
```
function generateUniqueId() {
    return (Date.now().toString(36) + Math.random().toString(36).substr(2, 5)).toUpperCase();
}
```

and then use as:
```
input.id = input.name = 'attachment_' + generateUniqueId();
```

File: blocks/form/file.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        const dataTransfer = new DataTransfer();
+        dataTransfer.items.add(file);
+        input.files = dataTransfer.files;
+        const div = document.createElement('div');
+        const remove = document.createElement('button');
+        remove.className = 'remove';
+        div.innerHTML = `<span>${file.name}</span>`;
```
Comment (solaris007) [inline blocks/form/file.js:12]: here we should potentially sanitize, albeit small-scale, against XSS. you could do:
```
function sanitizeHTML(input) {
    const map = {
        '&': '&amp;',
        '<': '&lt;',
        '>': '&gt;',
        '"': '&quot;',
        "'": '&#039;'
    };
    return input.replace(/[&<>"']/g, function(m) { return map[m]; });
}
```

and then use:
```
div.innerHTML = `<span>${sanitizeHTML(file.name)}</span>`;
```

File: blocks/form/file.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        button.innerText = 'Drop Here';
+        dropArea.innerHTML = '<p>Drop files here or</p>';
+        button.onclick = () => {
+            const input = fileInput.cloneNode(true);
+            input.multiple = true;
+            input.onchange = () => {
+                attachFiles(wrapper, input.files, fileInput);
+                input.remove();
+            }
+            input.click();
```
Comment (vdua) [inline blocks/form/file.js:44]: In the past fileInput that was not inserted in dom, didn't use to work in IE. Can you please confirm whether it is working on Edge/Firefox as well ?

File: blocks/form/file.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        const button = document.createElement('button');
+        button.type = 'button';
+        button.innerText = 'Drop Here';
+        dropArea.innerHTML = '<p>Drop files here or</p>';
+        button.onclick = () => {
+            const input = fileInput.cloneNode(true);
+            input.multiple = true;
+            input.onchange = () => {
+                attachFiles(wrapper, input.files, fileInput);
+                input.remove();
```
Comment (vdua) [inline blocks/form/file.js:42]: where are we inserting this input in the dom ?

File: blocks/form/file.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
+function attachFiles(wrapper, fileList, fileTemplate) {
+    for(let i = 0; i < fileList.length; i++) {
+        const file = fileList[i];
+        const input = fileTemplate.cloneNode(true);
+        input.id = input.name = 'attachment_' + Date.now() + '' + Math.floor((Math.random() * 1000));
+        if (input.accept.includes(file.type)) {
+            const dataTransfer = new DataTransfer();
+            dataTransfer.items.add(file);
+            input.files = dataTransfer.files;
```
Comment (vdua) [inline blocks/form/file.js:9]: I am surprised that this works.

Comment (Pareesh) [reply blocks/form/file.js:9]: Need to validate this usecase only with submit service.

File: blocks/form/file.js (hunk 2 above)
Comment (Pareesh) [reply blocks/form/file.js]: Removed the usage of generating unique, using only one input.

File: blocks/form/file.js (hunk 3 above)
Comment (Pareesh) [reply blocks/form/file.js]: using innertext instead of innerHTML, to avoid risk of XSS.

File: blocks/form/form.js [hunk 7]
```diff
@@ -202,6 +207,9 @@ const createSelect = withFieldWrapper((fd) => {
     const option = document.createElement('option');
     option.textContent = o.trim();
     option.value = o.trim();
+    if (fd.Value === o.trim()) {
+      option.selected = true;
```
Comment (jalagari) [inline blocks/form/form.js:211]: option.selected = fd?.Value === o?.trim()

File: blocks/form/file.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
+function isFileAllowed(file, allowedTypes) {
+  if (!file) {
+    return false;
+  }
+  const fileExtension = file.name.split('.').pop().toLowerCase();
```
Comment (jalagari) [inline blocks/form/file.js:5]: What if file doesn't have extension

File: blocks/form/file.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+function getFileList(files) {
+  const dataTransfer = new DataTransfer();
+  files.forEach((file) => dataTransfer.items.add(file));
+  return dataTransfer.files;
+}
+
+export default async function decorate(element) {
+  const wrappers = element.querySelectorAll('.form-browse');
```
Comment (jalagari) [inline blocks/form/file.js:17]: Lets add consistency behaviour for all file input. Contact us doesn't have drop box

**Comment by solaris007** [comment]
@Pareesh @jalagari please investigate the CLS and TBT before merging.

File: blocks/form/form.js [hunk 10]
```diff
@@ -20,12 +21,16 @@ function constructPayload(form) {
   const payload = {};
```
Comment (neerajadobe93) [inline blocks/form/form.js:21]: @Pareesh  payload should be initialized with unique id as in afb : https://github.com/adobe/afb/blob/main/blocks/form/form.js#L24

File: blocks/form/file.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
+function isFileAllowed(file, allowedTypes) {
+  if (!file) {
```
Comment (solaris007) [inline blocks/form/file.js:2]: Would throw here as this should not happen:
```
  if (!file) {
    throw new Error("File object is required.");
  }
```

File: blocks/form/form.js [hunk 12]
```diff
@@ -20,12 +21,16 @@ function constructPayload(form) {
...
   const attachments = {};
   [...form.elements].forEach((fe) => {
-    if (fe.type === 'checkbox') {
-      if (fe.checked) payload[fe.id] = fe.value;
-    } else if (fe.type === 'file' && fe.files?.length > 0) {
-      attachments[fe.name] = fe.files;
-    } else if (fe.id) {
-      payload[fe.id] = fe.value;
+    if (fe.name) {
```
Comment (solaris007) [inline blocks/form/form.js:24]: we could reduce the nesting level and repetition slightly:
```
[...form.elements]
  .filter(fe => fe.name) // Filtering out elements without names upfront
  .forEach((fe) => {
    if (fe.type === 'radio' || fe.type === 'checkbox') {
      // Handle radio and checkbox elements
      if (fe.checked) {
        payload[fe.name] = fe.type === 'checkbox' && payload[fe.name]
          ? `${payload[fe.name]}, ${fe.value}`
          : fe.value;
      }
    } else if (fe.type === 'file' && fe.files?.length > 0) {
      attachments[fe.name] = fe.files;
    } else {
      payload[fe.name] = fe.value;
    }
  });

```

File: blocks/form/file.js (hunk 9 above)
Comment (Pareesh) [reply blocks/form/file.js:17]: @jalagari you want me to add the dropbox in submit a tech support ticket form ??

[the remark this answers is not in the document]
Comment (vdua) [reply blocks/form/file.js:17]: It should be not based on the name of the field but based on field type

File: blocks/form/form.js (hunk 7 above)
Comment (vdua) [reply blocks/form/form.js:211]: I am not convinced with optional chaining since it doesn't work in older browsers

File: blocks/form/form.js [hunk 13]
```diff
@@ -221,6 +229,12 @@ const createOutput = withFieldWrapper((fd) => {
   return output;
 });
 
+const createFile = withFieldWrapper((fd) => {
+  const input = createInput(fd);
+  input.accept = fd.Accept || '';
```
Comment (vdua) [inline blocks/form/form.js:234]: do we allow attaching multiple files ?

File: blocks/form/form.js [hunk 14]
```diff
@@ -321,20 +336,20 @@ async function createForm(formURL) {
...
     const input = el.querySelector('input,textarea,select');
-    if (fd.Mandatory && fd.Mandatory.toLowerCase() === 'true') {
-      input.setAttribute('required', 'required');
-    }
     if (input) {
       input.id = fd.Id;
       input.name = fd.Name;
       input.value = fd.Value;
+      input[fd.Mandatory && fd.Mandatory.toLowerCase() === 'true'
+        ? 'setAttribute' : 'removeAttribute']('required', '');
```
Comment (vdua) [inline blocks/form/form.js:344]: nitpick: not readable and can be prone to error.

**Comment by solaris007** [comment]
[/us/en/about-us/innovate-with-mammotome](https://mammotomeformstyle--mammotome--hlxsites.hlx.page/us/en/about-us/innovate-with-mammotome) seems to have a high CLS. can you investigate please?

**Comment by Pareesh** [comment]
Hi @solaris007 ,
I have checked the CLS, this is not happening because of form block but due to `Fonts causing FOIT/FOUT`. Font loading cause the shifting of text appear over the hero image. 

I am not sure why but different PSI run is giving variable result for same changes. 

cc : @jalagari

**Review by dzehnder — CHANGES_REQUESTED** [review]
Please move the svg icons from the code to the SharePoint icon folder. This way, the authors will be able to change them without changing the code.

**Review by solaris007 — APPROVED** [review]

File: blocks/form/form.js [hunk 15]
```diff
@@ -16,16 +17,22 @@ function loadScript(url) {
   return script;
 }
 
+function generateUniqueId() {
+  return (Date.now().toString(36) + Math.random().toString(36).slice(2, 7)).toUpperCase();
+}
+
 function constructPayload(form) {
-  const payload = {};
+  const payload = { __id__: generateUniqueId() };
```
Comment (jalagari) [inline blocks/form/form.js:25]: This is no longer required as we are not using Power Automate

File: blocks/form/form.js (hunk 7 above)
Comment (jalagari) [reply blocks/form/form.js:213]: 94% coverage is there for optional https://caniuse.com/?search=optional%20chaining%20

File: styles/styles.css [hunk 16]
```diff
@@ -934,7 +906,13 @@ main .schedule-demo .default-content-wrapper h2 {
 
 /* Position preceding content */
 .section.arc-after-section {
-  margin-top: -20%;
+  margin-top: -13%;
+  position: relative;
+}
+
```
[the remark this answers is not in the document]
Comment (dzehnder) [reply styles/styles.css:912]: This should be fine in combination with #273. Veryfied just now

File: styles/styles.css [hunk 17]
```diff
@@ -934,7 +906,13 @@ main .schedule-demo .default-content-wrapper h2 {
...
+/* stylelint-disable-next-line no-descending-specificity */
+.section.arc-after-section p { 
+  text-align: left;
```
Comment (dzehnder) [inline styles/styles.css:915]: Please do not use `stylelint-disable` and move it to the corresponding location in the css

**Pull request merged**